Improve To-do service error handling

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update To-do service error handling to use ServiceValidationError and translations. These are now `HomeAssistantError` which allows `continue_on_error` to work.

Note: This is not a "new" problem that is a release blocker, so it can go with any upcoming patch milestone


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #105402
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

allenporter committed:
Improve To-do service error handling

edenhaus commented on homeassistant/components/todo/__init__.py:
```diff
@@ -106,7 +106,7 @@ def _validate_supported_features(
             continue
         if not supported_features or not supported_features & desc.required_feature:
-            raise ValueError(
+            raise ServiceValidationError(
```
It would be great if we make the errors translatable too.
https://developers.home-assistant.io/docs/internationalization/core/#exceptions

Do you think we can add it to this PR?

allenporter replied:
Sure, i'll have a look.

edenhaus commented:
@allenporter  `ServiceValidationError` is a subsclass of `HomeAssistantError` and should also work for `continue_on_error`. So I think we can use `ServiceValidationError` also for the other places

allenporter committed:
Add service validation translations and fix shopping list

allenporter committed:
Convert to ServiceValidationError for all cases and add more translations

allenporter commented:
Updated to use `ServiceValidationError` everywhere with translations, updated the tests (including shopping list) and manually tested everything is working:
<img width="1085" alt="Field not supported" src="https://github.com/home-assistant/core/assets/6026418/da625be2-17a2-4260-b33a-22016180e4ee">
<img width="1081" alt="Item not found" src="https://github.com/home-assistant/core/assets/6026418/5f73366f-af37-41f7-b000-e20990b9a027">

edenhaus reviewed (approved):
Thanks @allenporter :+1:

MartinHjelmare commented on homeassistant/components/todo/strings.json:
```diff
@@ -90,5 +90,13 @@
+  },
+  "exceptions": {
+    "item_not_found": {
+      "message": "Unable to find To-do item: {item}"
```
We call it `to-do list item` in the other strings above.
